Check for partially spent token if redeem fails

robwoodgate opened the pull request:
This PR will check and handle partially spent tokens if redeem fails.

<img width="1440" height="932" alt="Screenshot 2025-11-24 at 15 59 39" src="https://github.com/user-attachments/assets/143a4991-25e1-4d31-8cdd-656a0f7d8d92" />

robwoodgate commented:
@callebtc - have rebased to current main to make review easier.

robwoodgate commented:
Rebased again... please let me know if this contribution is a no go.
Assuming am ok to delete the [alternative branch.](https://github.com/cashubtc/cashu.me/pull/353)

lollerfirst commented:
I think this might be dangerous. If the intended amount promised to receive is not fulfilled, it doesn't make sense to claim only a portion of it.

Example: you give some banknotes to someone, but they reject some of the banknotes because they are ripped or otherwise invalid. If they claim the valid banknote but reject the payment, then you're gonna have a problem. Vice versa, they can't accept only part of the money and accept the payment.

robwoodgate commented:
@callebtc , @prusnak - rebased again. If this change isn't wanted, LMK as it's been open 10 months already.

robwoodgate committed:
Retry receive with unspent proofs if token partially spent

If the mint rejects a receive with "Token already spent", check the
proof states and retry with just the unspent ones. Uses the new
groupProofsByState() helper from cashu-ts 4.3.0.

robwoodgate commented:
Rebased and updated for CTS v4.3.0

robwoodgate committed:
Preserve token unit when rebuilding partially spent token

cashu-ts defaults a missing unit to "sat", so re-encoding a non-sat
token without it caused the retry to fail with "Token is not in wallet
unit".

robwoodgate committed:
Don't mutate receive store when rebuilding partial token

Pass the retry token to runReceive as a parameter so the input field
keeps showing what the user pasted if the retry fails.
